### PR TITLE
[precompile] Capture across calls the caller makes, not calls precompile makes

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #195284
* #195215
* __->__ #195070
* #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194452
* #194450
* #194449
* #194428
* #194427
* #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* #192379
* #193976
* #193697
* #192374
* #192866

precompile() runs every entry of example_inputs itself, back to back, with
nothing in between. That is only correct when the calls are independent, and a
pipelined training step is not: its inputs come off a queue that the enclosing
loop advances, so the second call finds the state the first one consumed. On a
real sharded recommendation model that surfaces as "Invalid PipelinedForward
usage, input_dist ... probably consumed by others" -- and no shape where
precompile drives the iteration fixes it, generator or callback included,
because advancing the queue is caller code outside the region.

accumulate() inverts the control: the caller keeps their loop.

    with torch.compiler.precompile.accumulate(
        train_step, artifact_path="m.py", cache_path="m.cache", training=True,
    ) as capture:
        for batch in loader:
            losses = capture(model, batch)
            optimizer.step()

Each call runs for real, folds in whatever it newly exercised, rewrites both
files, and returns what the region returned. A call that exercises nothing new
adds nothing. No finalize step: the files are a loadable artifact for everything
captured so far from the first call onwards, so a job that dies partway leaves a
working artifact for the batches it reached.

A handle rather than a state value threaded through precompile(). Not a style
preference -- the state cannot be a value. Compiled variants are filed under an
isolate_recompiles_id, lookup matches it with no fallback, and nothing can hand
an existing id to torch._dynamo.optimize: it mints its own from a private
counter. So a later call can only see an earlier call's variants if the compiled
region is still alive, which means the state IS a live object. Spelling it as a
keyword argument would have hidden that. Carrying the configuration on the
handle also makes a mid-capture disagreement about backend or training
unrepresentable rather than an error to raise.

The correctness problem this had to solve first: the invariant-guard policy
CANNOT be re-applied per call. _apply_guard_policy rewrites each variant's
serialized guards and then prunes the facts it derived them from, so a second
pass reads bytes the first already filtered against facts the first already
edited, and filtering only ever removes. That is not a slow leak. A single
variant is invariant by construction -- one variant cannot disagree with itself,
so varying_guard_slots over it is empty -- and the pass after call 1 therefore
strips every at-risk slot of every frame, permanently. When a later call adds
the variant that makes one of them discriminate, the earlier variant no longer
carries it and goes on matching calls that belong to the new graph. Measured end
to end on a HASATTR branch: the artifact serves the first call's numbers for the
second call's input.

So the policy is split. _apply_guard_policy still consumes the session, which is
right for the one-shot path where the capture is over. _policy_filtered_codes
runs the same core over deep copies and leaves the session pristine, and that is
what the accumulating render uses. deepcopy treats bytes as atomic, so it
duplicates the record structure and not the serialized payloads.

Resuming a session turned up two more things it could not have done before.
_closing is set by every exit and, on the one-shot path, never read again; a
second entry refused every call with "PrecompileSession is not active", and the
error was swallowed into _capture_errors where it failed require_complete
instead of being raised. And record_live_guard_leaves rebinds a fresh set on
each entry, so keeping only the newest cycle's leaves made _report_guard_drift
call every frame the cycle had not recompiled drifted; the leaves are unioned
across cycles now.

Cost is linear in what has been captured, not in the call: ~2.7 ms per
accumulated guarded code and ~10 KB of rendered source per variant, measured
over twelve calls. A long loop over a large model pays that every call, which is
the price of the files always being valid.

Three things a review pass caught in the first version of this, each verified
before and after:

The rewrite was not atomic. _write_artifact truncated both files where they lay,
code first then cache, and the pair only loads together -- the cache carries a
sha256 of exactly the python_code it was emitted with. So a process that died
mid-write left a new artifact beside the previous cache, which refuses to load
with "cache does not match python_code". At hundreds of megabytes that window is
seconds long on every call, which is exactly the crash the always-valid promise
is selling protection against. Both halves are written beside their targets,
fsynced, and renamed into place; a failed write leaves the previous pair intact
and no temp files behind.

A call that raised poisoned every later one. require_complete gates on
summary().capture_errors, which accumulates for the life of the session, so
after one bad call every subsequent render refused -- and refused AFTER the
region had run, so a training step's gradients moved and its result was never
returned. The gate now sees only the errors of the cycle it is rendering; a call
that failed has already raised to the caller.

The artifact under-reported its own drops. _policy_filtered_codes discarded what
the pass dropped, and _apply_guard_policy never runs while resumable, so
POLICY_DROPPED_GUARDS in the written file was empty where the one-shot path
listed AUTOGRAD_SAVED_TENSORS_HOOKS, DEFAULT_DEVICE, GRAD_MODE and the rest. The
drops are read back off the pass that made them, and REPLACED rather than
unioned, because keep_only grows as calls add variants and a slot dropped by an
earlier render can survive a later one. kept_guards is derived by subtraction
for the same reason instead of being decremented in place.

Test Plan:

```
python test/test_precompile.py
python test/dynamo/test_package.py
python test/dynamo/test_guard_serialization.py
python test/functorch/test_compile_to_python.py
```

Nine new tests. The one that matters is
test_precompile_accumulate_serves_a_slot_only_a_later_call_makes_vary: it builds
the HASATTR shape above, captures the two branches on two separate calls, and
asserts the reloaded artifact serves BOTH -- it returns the first branch twice
if the policy is applied to the capture rather than to a copy. The others cover
accumulation and result pass-through, gradients matching eager, the artifact
being loadable mid-loop, the session's facts surviving a render, close()
semantics, and the make_fx and lone-path rejections. Three more cover the fixes
above: a write that fails partway leaves the previous pair loadable, a call
after a raising one renders and returns, and the accumulating artifact reports
the same policy drops the one-shot one does.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by the canonical precompile stack #195353-#195362; this change lives in #195362 (with the guard-drop policy half in #195360). Closing to consolidate review there. The branch is kept on GitHub.
